lib/nxt_state_machine/state_machine.rb

```diff
@@ -0,0 +1,197 @@
+module NxtStateMachine
+  class StateMachine
+    def initialize(name, class_context, event_registry, **opts)
+      @name = name
+      @class_context = class_context
+      @options = opts
+
+      @states = NxtStateMachine::StateRegistry.new
+      @transitions = []
+      @events = event_registry
+      @callbacks = CallbackRegistry.new
+      @error_callback_registry = ErrorCallbackRegistry.new
+      @defuse_registry = DefuseRegistry.new
+
+      @initial_state = nil
+    end
+
+    attr_reader :class_context, :events, :options, :callbacks, :name, :error_callback_registry, :defuse_registry
+    attr_accessor :initial_state
+
+    def get_state_with(method = nil, &block)
+      @get_state_with ||= (method || block) || raise_missing_configuration_error(:get_state_with)
+    end
+
+    def set_state_with(method = nil, &block)
+      @set_state_with ||= (method || block) || raise_missing_configuration_error(:set_state_with)
+    end
+
+    def set_state_with!(method = nil, &block)
+      @set_state_with_bang ||= (method || block) || raise_missing_configuration_error(:set_state_with!)
+    end
+
+    def state(*names, **opts, &block)
+      defaults = { initial: false }
+      opts.reverse_merge!(defaults)
+      machine = self
+
+      Array(names).map do |name|
+        if opts.fetch(:initial) && initial_state.present?
+          raise NxtStateMachine::Errors::InitialStateAlreadyDefined, ":#{initial_state.enum} was already defined as the initial state"
+        else
+          state = new_state_class(&block).new(name, self, **opts.reverse_merge(index: states.size))
+          states.register(name, state)
+          self.initial_state = state if opts.fetch(:initial)
+
+          class_context.define_method "#{name}?" do
+            machine.current_state_name(self) == name
+          end
+
+          state
+        end
+      end
+    end
+
+    def states(*names, **opts, &block)
+      # method overloading in ruby ;-)
+      return @states unless names.any?
```

Comment:
Personally I find negations or double negations more difficult to process 😅
```suggestion
      return @states if names.empty?
```